### PR TITLE
[ios] use python script to generate extension safe frameworks and code sign them

### PR DESCRIPTION
**Pull request by cyanglaz**
Generates the fat framework for extension safe variants and add them to the code sign list.

Also reverts part of https://github.com/flutter/engine/pull/45664 where the variants are uploaded in separate archive. Now with this change, the extension safe is inside a "extension_safe" folder under existing archive. 

fixes https://github.com/flutter/flutter/issues/124291

The destination folder after this change:
- An extension_safe folder is added, which includes the extension safe Flutter.frameworks and dsym
- The artifacts.zip now includes the `extension_safe` except the dysm
- A extension_safe_Flutter.dsym.zip is created (under ios/release/) , which includes the extension_safe/Flutter.dsym

Screenshot 2023-09-13 at 3 26 23 PM

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: ci/builders/mac_ios_engine.json [hunk 1]
```diff
@@ -378,24 +331,9 @@
             "destination": "ios-release/Flutter.dSYM.zip",
             "realm": "production"
         },
-        {
```
Comment (godofredoc) [inline ci/builders/mac_ios_engine.json:381]: No needed anymore?

File: sky/tools/create_full_ios_framework.py [hunk 2]
```diff
@@ -106,14 +106,59 @@ def main():
...
+  # build extension safe frameworks
+  extension_safe_dst = os.path.join(dst, 'extension_safe')
+  extension_safe_framework = os.path.join(
+      extension_safe_dst, 'Flutter.framework'
+  )
+  extension_safe_simulator_framework = os.path.join(
+      extension_safe_dst, 'sim', 'Flutter.framework'
+  )
+  extension_safe_arm64_framework = os.path.join(
+      arm64_out_dir + '_extension_safe', 'Flutter.framework'
```
Comment (godofredoc) [inline sky/tools/create_full_ios_framework.py:119]: Here and below  '%s_extension_safe' % arm64_out_dir?

File: sky/tools/create_full_ios_framework.py [hunk 3]
```diff
@@ -106,14 +106,59 @@ def main():
       args, dst, framework, arm64_framework, simulator_framework,
       simulator_x64_framework, simulator_arm64_framework
   )
+
+  # build extension safe frameworks
```
Comment (godofredoc) [inline sky/tools/create_full_ios_framework.py:110]: Can we move this part its own function?

File: ci/builders/mac_ios_engine.json (hunk 1 above)
Comment (cyanglaz) [reply ci/builders/mac_ios_engine.json:381]: Yeah, so i moved the Flutter.framework for extension_safe into the responding existing zips. In below python change
For example
debug extension_safe Flutter.framework is now moved under ios-debug/extension_safe

File: sky/tools/create_full_ios_framework.py [hunk 4]
```diff
@@ -106,6 +106,46 @@ def main():
...
+  extension_safe_dst = os.path.join(dst, 'extension_safe')
```
Comment (cyanglaz) [inline sky/tools/create_full_ios_framework.py:111]: Here, I created a new subfolder under dst, extension safe frameworks are all moved into this subfolder.

**Comment by cyanglaz** [comment]
@godofredoc Updated per review comments. I also updated the PR description to reflect file changes.

**Comment by stuartmorgan** [comment]
> * A extension_safe_Flutter.dsym.zip is created, which includes the extension_safe/Flutter.dsym

Why is this next to the directory rather than in it? It seems inconsistent with both the other extension-safe artifacts and the main framework's dSYM handling.

**Comment by cyanglaz** [comment]
>Why is this next to the directory rather than in it? It seems inconsistent with both the other extension-safe artifacts and the main framework's dSYM handling. 

The main Flutter.dsym is not included in the artifact.zip. So I thought it is more consistent to put the extension_safe one outside of the artifact.zip too.

File: sky/tools/create_full_ios_framework.py [hunk 5]
```diff
@@ -192,12 +233,19 @@ def zip_archive(dst):
       'Flutter.xcframework',
       'entitlements.txt',
       'without_entitlements.txt',
+      'extension_safe/Flutter.framework',
+      'extension_safe/Flutter.xcframework',
+      'extension_safe/sim',
   ],
                         cwd=dst)
   if os.path.exists(os.path.join(dst, 'Flutter.dSYM')):
```
Comment (cyanglaz) [inline sky/tools/create_full_ios_framework.py:241]: We explictly put the Flutter.dsym in a standalone zip file. So I thought we should do the same for extension safe dsym to keep it consistent

Comment (cyanglaz) [reply sky/tools/create_full_ios_framework.py:241]: Oh I think this is what is misleading. I don't think the zip file location mattered, they are not really useful here, they are later uploaded to a destination determined in the mac_ios_engine.json. After my change, both Flutter.dsym and extension_safe_Flutter.dsym are uploaded to ios/release, where Flutter.dsym used to be.

**Review by stuartmorgan — COMMENTED** [review]
Looks reasonable to me, but I'm not very familiar with this iteration of the builder scripts so someone more familiar with them should do the main review.

**Pull request merged**